### ISSUE TITLE
revise only saves updated data on the second run

### ISSUE THREAD

**Issue by chris-aeviator** [telepathylabsai/OpenDF#29]
I'm trying to call `revise` to update an `Event`

Therefore I consider the following dialog

```
CreateEvent(AND(has_subject(AI Training for our website),with_attendee(dan),starts_at(AND(Today(),HourMinuteAm(hours=11,minutes=30)))))

Does this look ok? -
AI Training for our website with 2 Attendees
Dan Smith
You: notresponded, the status is busy
starting today at 11:30 lasting 30 minutes
and held online

AcceptSuggestion()

AI Training for our website with 2 Attendees NL Dan Smith: notresponded, the status is busy NL You: notresponded, the status is busy NL starting today at 11:30 lasting 30 minutes NL and held online

revise(old=Event??(), new=0golf, inp_nm=subject, newMode=extend)

AI Training for our website with You: notresponded, the status is busy starting today at 11:30 lasting 30 minutes NL and held online
```

![grafik](https://user-images.githubusercontent.com/11522213/226931757-5f38794b-c03b-4880-a98b-2c7c19706831.png)

as you can see the event name has not been updated (same in the graphViz graph)

However, if I run the command again, this time with a different subject, I can see my previous command  (`subject=0golf`) being evaluated. 

![grafik](https://user-images.githubusercontent.com/11522213/226932480-12e4a5ae-aa12-49c5-a625-ba1f3796c075.png)

I can confirm that also the database (`evt.add_event` call) only receives the correct value on the second `revise` call, with the data being from the 2nd previous call

EDIT:
I'm using `resources/smcalflow_config.yaml`

![grafik](https://user-images.githubusercontent.com/11522213/226941105-a50a8d76-1a18-4841-8001-6cacc472467b.png)

**Comment by guimaraes-tl** [comment]
Hello chris-aeviator,

Thank you for your comment.

### TL;DR

The fix for what you asked is to use the revise expression below:
`revise(old=Event??(), new=0golf, hasInput=subject, newMode=extend)`

However, what you might want to do is:
```
UpdateEvent(event=has_subject(AI Training for our website), constraint=has_subject(0golf))
AcceptSuggestion()
```
## More detailed answer

When you run the expression: `CreateEvent(AND(has_subject(AI Training for our website),with_attendee(dan),starts_at(AND(Today(),HourMinuteAm(hours=11,minutes=30)))))` you are creating a Dataflow (DF) graph with a set of constraints that will define the fields of the event you want to create. In this case, `subject=AI Training for our website`, `attendee=dan`, etc...

These constraints will be given to an event factory that will try to come up with a **concrete** (fully specified) event that satisfies the constraints. If the factory **is not able** to create such event, DF will ask the user for more constraints (or to modify existing ones), until the factory is able to create the concrete event.

Assuming it was able to create the concrete example, it will finally ask the user to confirm the event. After the confirmation by the user, which is given by `AcceptSuggestion()`, the event is stored in the database.

When you use `revise(old=Event??(), new=0golf, inp_nm=subject, newMode=extend)` you are not modifying the event itself, you are modifying the constraints used by the factory to create the event. This is denoted by the double `?` on `old=Event??()`.

The problem here is that `revise(old=Event??(), new=0golf, inp_nm=subject, newMode=extend)` is revising the wrong constraint, in the set of constraints. This results in adding another constraint saying that the subject of the event should be `0golf`, but the previous constraint saying that the name of the event should be `AI Training for our website` still exists.

When the constraints are passed to the event factory, some constraints might contradict each other (e.g. a constraint saying that the subject must be A and another saying that the subject must be B). To resolve this, a pruning mechanism discard incompatible constraints based on some heuristics. In this case, it is discarding the `0golf` subject.

During the second revise `revise(old=Event??(), new=XXXXXX, inp_nm=subject, newMode=extend)`, the pruning mechanism manages to consider the desired subject `0golf`. Note that the full set of constraint is still in the graph, the pruning mechanism creates a copy of the constraint every time, without changing the original ones.

You can solve this particular issue, but **it still might not be what you really want**.
To fix this issue, you could use `hasInput` instead of `inp_nm`. Resulting in: `revise(old=Event??(), new=0golf, hasInput=subject, newMode=extend)`. This will replace the old subject constraint directly, and there will be no more conflicting subject constraints to be pruned.

However, here you are revising the constraint that generated the event, not the event itself. And even if you change the revise expression to revise the event directly, you would revise the DF graph representation of the example.

In a real application, I assume you actually want to revise the event stored in the database, which is the part persistent to the user (possibly also to other users as well), not the DF graph itself, which is only the context of the current conversation.

If you revise the event constraint directly, it will result in a new execution of the `CreateEvent` node, which will create a new event in the database, keeping the first version of the event unchanged. To proper change an event **after** confirming its creation, one should use `UpdateEvent` instead. In this case, one could use: `UpdateEvent(event=has_subject(AI Training for our website), constraint=has_subject(0golf))`. Then, use `AcceptSuggestion()` to confirm the change.

If you want to confirm the duplication of event in the database, you can uncomment the line `222` of the file `opendf/defs.py` and comment the line `223`. This will make DF to use a sqlite file database in the path `tmp/test.db`, the default is a memory database. Then, you can manually inspect it using your preferred database tool. The name of the table is `event`.

Best regards.

**Comment by chris-aeviator** [comment]
this worked like a charm - thank you very much for the very detailed information. I'm very interested in continuing in using openDF for neurosymbolic AI and extremely happy seeing such a good ownership of a repo :+1:
